tests/src/main/java/net/neoforged/neoforge/debug/client/ClientEventTests.java

```diff
@@ -126,4 +140,47 @@ static void renderPlayerEvent(final DynamicTest test) {
             test.requestConfirmation(Minecraft.getInstance().player, Component.literal("Is an iron block rendered above you in third-person?"));
         });
     }
+
+    @TestHolder(description = { "Test render state modifier system and registration event" })
+    static void updateRenderState(final DynamicTest test) {
+        var rotationKey = new ContextKey<Float>(ResourceLocation.fromNamespaceAndPath(test.createModId(), "rotation"));
+        var numRenderAttachmentKey = new ContextKey<Integer>(ResourceLocation.fromNamespaceAndPath(test.createModId(), "times_to_render"));
+        var testAttachment = test.registrationHelper().attachments().registerSimpleAttachment("test", () -> 3);
+        test.whenEnabled(listeners -> {
+            listeners.mod().addListener((RegisterRenderStateModifiersEvent event) -> {
+                event.registerEntityModifier(PlayerRenderer.class, (entity, renderState) -> {
+                    renderState.setRenderData(rotationKey, 45f);
+                });
+                event.registerEntityModifier(new TypeToken<LivingEntityRenderer<? extends LivingEntity, LivingEntityRenderState, ?>>() {}, (entity, renderState) -> {
+                    renderState.setRenderData(numRenderAttachmentKey, entity.getData(testAttachment));
+                });
+                // Test other type parameters for safety
+                event.registerEntityModifier(new TypeToken<AbstractHoglinRenderer<?>>() {}, (entity, renderState) -> {});
+                event.registerEntityModifier(new TypeToken<MobRenderer<Mob, LivingEntityRenderState, ?>>() {}, (entity, renderState) -> {});
+                try {
+                    class TestBrokenHoglinRendererTypeToken<T extends Mob & HoglinBase> extends TypeToken<AbstractHoglinRenderer<T>> {}
+                    event.registerEntityModifier(new TestBrokenHoglinRendererTypeToken<>(), (entity, renderState) -> {});
+                    test.fail("Unsafe type parameter succeeded. Cannot assume T can be ?.");
+                } catch (IllegalArgumentException ignored) {}
+            });
```

Comment:
The modifier registration must be moved outside the `test.whenEnabled()` lambda and registered via `test.framework().modEventBus()` instead, otherwise the modifiers won't be present when the test is enabled after joining a world, causing the test to fail.